[LIVY-551] Add "doAs" impersonation support

risdenk opened the pull request:
## What changes were proposed in this pull request?

Currently `proxyuser` is used in batches and sessions for impersonation. This should be extended to support impersonation for all endpoints. Adding the query parameter `doAs` matches others in the Hadoop ecosystem and will work across endpoints.

https://issues.apache.org/jira/browse/LIVY-551

## How was this patch tested?

Added unit and integration tests for this feature. Manually tested to ensure that `proxyuser` still works. Also checked that if both `proxyuser` and the new `doAs` parameter are specified that `doAs` takes precedence.

risdenk commented:
Most of the changes are to pass the Session and HttpServletRequest around to allow for checking different parts of the request with the access manager.

risdenk commented:
The latest build succeeded except for "Spark 2.3 Unit Tests"

```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.livy.rsc.TestJobHandle
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.634 sec - in org.apache.livy.rsc.TestJobHandle
Running org.apache.livy.rsc.TestSparkClient
Tests run: 19, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 57.489 sec <<< FAILURE! - in org.apache.livy.rsc.TestSparkClient
testRemoteClient(org.apache.livy.rsc.TestSparkClient)  Time elapsed: 4.546 sec  <<< ERROR!
java.util.concurrent.ExecutionException: java.io.IOException: RSCClient instance stopped.
	at org.apache.livy.rsc.TestSparkClient.runTest(TestSparkClient.java:588)
	at org.apache.livy.rsc.TestSparkClient.runTest(TestSparkClient.java:575)
	at org.apache.livy.rsc.TestSparkClient.testRemoteClient(TestSparkClient.java:213)
Caused by: java.io.IOException: RSCClient instance stopped.

Running org.apache.livy.rsc.rpc.TestKryoMessageCodec
Tests run: 9, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.042 sec - in org.apache.livy.rsc.rpc.TestKryoMessageCodec
Running org.apache.livy.rsc.rpc.TestRpc
Tests run: 7, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 0.407 sec - in org.apache.livy.rsc.rpc.TestRpc

Results :

Tests in error: 
  TestSparkClient.testRemoteClient:213->runTest:575->runTest:588 Â» Execution jav...
```

None of the other Spark versions failed. I can't rekick off the single job since it looks to be a transient failure.

risdenk commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -97,47 +99,71 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
+  def checkProxyUser(request: HttpServletRequest, impersonatedUser: String): Boolean = {
+    val proxyUser = getRequestUser(request)
+    val remoteHost = request.getRemoteHost
+    val allowedHosts = conf.hadoopConf.get("hadoop.proxyuser." + proxyUser + ".hosts")
```
The idea was to model after https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Superusers.html#Configurations and how a bunch of other Hadoop ecosystem components work.

Superusers can impersonate anyone yes. There should be a limit as to who can impersonate who. There should also be a limit from where superuser requests can come from.

vanzin replied:
If you want to add this functionality, you shouldn't use the Hadoop configs. Every different service has its own configs for this.

It should also be a separate change, since it's not really related to what you're implementing here.

risdenk replied:
Yea thats a valid point that it should be a separate change.

risdenk commented:
Pushed updated changes that:
* removes the Hadoop proxyuser configs 
* fixes formatting for parameters in multi-line declarations
* Renamed `checkBodyProxyUser` method to explain better what it does
* Added/improved comments to new/changed methods

vanzin commented on server/src/test/scala/org/apache/livy/server/AccessManagerSuite.scala:
```diff
@@ -116,4 +121,60 @@ class AccessManagerSuite extends FunSuite with Matchers with LivyBaseUnitTestSui
+    when(req.getParameter("doAs")).thenReturn("def")
+
+    accessManager.getRequestUser(req) should be("abc")
+    try {
```
```
val e = intercept[AccessControlException] {
  // something that throws
}
```

vanzin commented on server/src/test/scala/org/apache/livy/server/AccessManagerSuite.scala:
```diff
@@ -116,4 +121,60 @@ class AccessManagerSuite extends FunSuite with Matchers with LivyBaseUnitTestSui
+
+    accessManager.getRequestUser(req) should be("abc")
+    try {
+      accessManager.getImpersonatedUser(req) should be(Some("def"))
```
the "should be" doesn't seem necessary since you expect an exception here. Also in the next test.

risdenk commented:
* Added tests to `SessionServletSpec` and `AclsEnabledSessionServletSpec`
* Cleaned up test error handled to use `intercept[...]`
* removed unnecessary `should be` usage

mgaido91 commented on server/src/main/scala/org/apache/livy/server/batch/BatchSessionServlet.scala:
```diff
@@ -42,20 +42,24 @@ class BatchSessionServlet(
     val createRequest = bodyAs[CreateBatchRequest](req)
+    val owner = accessManager.getRequestUser(req)
+    val legacyProxyUser = getLegacyProxyUser(owner, createRequest.proxyUser, req)
+    val impersonatedUser = getImpersonatedUser(req).orElse(legacyProxyUser)
```
we are doing the parsing for `legacyProxyUser ` also when it is not needed, we can do:
```
orElse(getLegacyProxyUser(....))
```

risdenk replied:
Addressed in latest commit.

risdenk commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +100,79 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
-   *         disabled.
+   * Gets the remote user from the given request
+   */
+  def getRequestUser(request: HttpServletRequest): String = {
```
Logic wasn't bad to move out of AccessManager so took that approach in latest commit.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala:
```diff
@@ -55,12 +54,11 @@ object BatchSession extends Logging {
       sessionStore: SessionStore,
       mockApp: Option[SparkApp] = None): BatchSession = {
     val appTag = s"livy-batch-$id-${Random.alphanumeric.take(8).mkString}"
-    val impersonatedUser = accessManager.checkImpersonation(request.proxyUser, owner, livyConf)
```
I am not sure about this. I feel it is a bit risky to make this check outside here, because we do not guarantee anymore that we create the session only if the impersonated user has the right to do so. Can't we leave it here?

vanzin replied:
It's just moving the check to the servlet, which happens before this is called. Are you worried that another call site might be added and we forget about the access check?

mgaido91 replied:
yes, exactly, that's my worry.

risdenk replied:
Ok I put the accessManager back

vanzin commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,51 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
   def checkImpersonation(
-      target: Option[String],
       requestUser: String,
-      livyConf: LivyConf): Option[String] = {
```
I don't really see a use case for the conf instances being different. At the call sites, it's just a matter of sub-optimal APIs (e.g. `BatchSession.create` being a static method and thus not having any state associated, like a configuration object).

vanzin commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -156,9 +156,39 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
-  protected def remoteUser(req: HttpServletRequest): String = req.getRemoteUser()
+    * Gets the remote user from the given request
+    */
+  protected def getRemoteUser(request: HttpServletRequest): String = {
```
Also I prefer the old naming style, without the java bean-esque "get", for the new methods below too.

risdenk replied:
Yea I can change it back. This happened from the refactor from accessManager back to SessionServlet.

risdenk commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,51 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
+  def hasModifyAccess(
+      session: Session,
+      effectiveUser: String): Boolean = {
+    session.owner == effectiveUser  ||
```
If you use a superuser to create a session, the owner is set to the superuser and proxyUser is the user running the job. If you try to access Livy directly as the user running the job, you won't be able to see details. This is weird since you should be able to see your own job. This just ass the check to see if the incoming user is either the owner or the proxyUser of the session.

mgaido91 replied:
Yes, but IIUC this is a behavior change. We can argue whether it is right or not the current behavior and eventually change it, but we should definitely not do that in this PR, as it is unrelated with it. If I am not missing something, please revert these changes and create a separate ticket in order to discuss them separately.

risdenk replied:
Changes reverted in latest push

risdenk commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,51 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
+  def hasViewAccess(
+      session: Session,
+      effectiveUser: String): Boolean = {
+    session.owner == effectiveUser ||
```
If you use a superuser to create a session, the owner is set to the superuser and proxyUser is the user running the job. If you try to access Livy directly as the user running the job, you won't be able to see details. This is weird since you should be able to see your own job. This just ass the check to see if the incoming user is either the owner or the proxyUser of the session.

risdenk commented:
Ok I think I addressed all the review comments:
* removed get from method names
* added accessManager back to session create() methods
* Removed checking of proxyuser in accessmanager for permissions
* fixed minor formatting issues

mgaido91 commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,43 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
-   *         disabled.
+   * Checks that the request user can impersonate the target user.
+   * If impersonation is enabled and the user does not have permission to impersonate
+   * then throws an `AccessControlException`. If impersonation is disabled returns false
```
nit: missing dot

mgaido91 commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,43 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
 
   /**
-   * Check that the requesting user has admin access to resources owned by the given target user.
+   * Check that the request user has is able to impersonate the given user.
```
the new description is wrong IMHO, and these are unneeded changes (just renaming variables). Please avoid them

mgaido91 commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,46 +99,43 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
 
   /**
-   * Check that the request's user has view access to resources owned by the given target user.
+   * Check that the request user has view access to the given session
```
nit: missing dot

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -156,9 +156,29 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
-   * Returns the remote user for the given request. Separate method so that tests can override it.
-   */
-  protected def remoteUser(req: HttpServletRequest): String = req.getRemoteUser()
+    * Returns the remote user for the given request. Separate method so that tests can override it.
```
bad indentation and unneeded changes, please avoid them

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -156,9 +156,29 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
+  }
+
+  /**
+    * Returns the proxy user as determined by the json request body or owner if available.
```
nit: bad indent

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -156,9 +156,29 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
+  }
+
+  /**
+    * Gets the request user or impersonated user to determine the effective user
```
ditto

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -183,11 +203,11 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
     sessionManager.get(sessionId) match {
       case Some(session) =>
-        if (allowAll || checkFn.map(_(session.owner, remoteUser(request))).getOrElse(false)) {
+        if (allowAll || checkFn.map(_(session, effectiveUser(request))).getOrElse(false)) {
```
again, this is a behavior change. Probably it is fine to leave it here, but at least we need to document this behavior change. cc @vanzin

risdenk replied:
If you only ever look at the remoteUser you can't authorize any other types of requests. That is the whole point of this PR that you can authorize not just POST create session/batch requests. This is described in the description of LIVY-551. This change is required.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala:
```diff
@@ -57,10 +57,12 @@ object BatchSession extends Logging {
     val appTag = s"livy-batch-$id-${Random.alphanumeric.take(8).mkString}"
-    val impersonatedUser = accessManager.checkImpersonation(request.proxyUser, owner, livyConf)
+
+    impersonatedUser.filter(accessManager.checkImpersonation(owner, _))
```
the result of the filter is never used, so it is useless as it is now... Moreover it is weird that tests are passing nonetheless... may you please add a UT for this?

risdenk replied:
Tests are passing because BatchSession/InteractiveSession create are never called directly. This was defensively put in based on your other comments. I reverted the change to remove it based on your request.

Having a test that fails on a PR is pointless. I don't understand what this comment of yours is even after.

Finally the result of the call doesn't actually matter since there will be an exception thrown by checkImpersonation if there is a problem. Otherwise security is turned off. There is no reason to filter the impersonated user. It is just cleaner than doing an isDefined check and a .get.

mgaido91 replied:
well, then `foreach` is better than `filter` if we return nothing. Anyway, it actually works because you are doing the same check in `def impersonatedUser` and `def legacyProxyUser`, so here you are relying on them, rather than on this check .What I meant was doing the check here as it was earlier and in those methods just check which user needs to be impersonated. I see that you need that for the `effectiveUser` call, but I think we can move the filter there and use this one here.

risdenk replied:
I think the latest changes in 7829cd9 are what you were after.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/batch/BatchSessionServlet.scala:
```diff
@@ -42,20 +42,25 @@ class BatchSessionServlet(
   override protected def createSession(req: HttpServletRequest): BatchSession = {
     val createRequest = bodyAs[CreateBatchRequest](req)
+    val owner = remoteUser(req)
+    val proxyUser = impersonatedUser(req)
```
this code is duplicated, what about creating a method `proxyUser`?

mgaido91 commented on server/src/main/scala/org/apache/livy/server/interactive/InteractiveSession.scala:
```diff
@@ -67,14 +67,16 @@ object InteractiveSession extends Logging {
     val appTag = s"livy-session-$id-${Random.alphanumeric.take(8).mkString}"
-    val impersonatedUser = accessManager.checkImpersonation(request.proxyUser, owner, livyConf)
+
+    impersonatedUser.filter(accessManager.checkImpersonation(owner, _))
```
ditto

mgaido91 commented on server/src/main/scala/org/apache/livy/sessions/Session.scala:
```diff
@@ -204,7 +204,7 @@ abstract class Session(val id: Int, val owner: String, val livyConf: LivyConf)
   protected def stopSession(): Unit
 
-  protected val proxyUser: Option[String]
+  val proxyUser: Option[String]
```
please add a comment:
```
// Visible for testing.
```

mgaido91 commented on server/src/test/scala/org/apache/livy/server/BaseSessionServletSpec.scala:
```diff
@@ -77,7 +77,6 @@ abstract class BaseSessionServletSpec[S <: Session, R <: RecoveryMetadata]
   addServlet(servlet, "/*")
 
   protected def toJson(msg: AnyRef): Array[Byte] = mapper.writeValueAsBytes(msg)
-
```
in general we best avoid doing unneeded changes like these, makes easier checking the history of the changes.. so please undo the changes.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -160,6 +160,30 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
+  protected def effectiveUser(request: HttpServletRequest): String = {
+    val requestUser = remoteUser(request)
+    impersonatedUser(request).orElse(Option(requestUser))
+      .filter(accessManager.checkImpersonation(requestUser, _)).orNull
```
I think this should be:
```
impersonatedUser(request).filter(accessManager.checkImpersonation(requestUser, _)).getOrElse(requestUser)
```
Am I wrong? Why are you checking the impersonation also when there is no impersonated user?

risdenk replied:
I think you are right here.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/batch/BatchSession.scala:
```diff
@@ -57,10 +57,12 @@ object BatchSession extends Logging {
     val appTag = s"livy-batch-$id-${Random.alphanumeric.take(8).mkString}"
-    val impersonatedUser = accessManager.checkImpersonation(request.proxyUser, owner, livyConf)
+
+    impersonatedUser.foreach(accessManager.checkImpersonation(owner, _))
```
IIUC, as the code is now, when impersonation is disabled and we are using the old proxyUser feature, we would allow the impersonation, right? Because having the foreach here doesn't throw any exception and the returned false is just ignored.

Anyway, can we add a UT for this case, in order to ensure the behavior is fine?

risdenk replied:
Yea so there are no test suites that previously test the no impersonation code path. `BatchServletSpec` and `InteractiveServletSpec` both mock the Session `create()` method so there is no tests there. `JobApiSpec` is the closest to a real test but that one uses impersonation enabled always with no easy way to override that config. Let me look into it a bit.

risdenk replied:
ebf3514 fixes the test failures from the added UT for this behavior. The old behavior is now restored and confirmed via tests.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -160,6 +160,30 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
+   */
+  protected def proxyUser(request: HttpServletRequest,
+      createRequestProxyUser: Option[String]): Option[String] = {
+    impersonatedUser(request).orElse(createRequestProxyUser).orElse(Option(remoteUser(request)))
```
why `orElse(Option(remoteUser(request)))`? In that case we should return `None` here

risdenk replied:
So this caused a failure in one of the existing test cases if you don't have the remoteUser set as the proxyUser. 

https://github.com/apache/incubator-livy/blob/master/server/src/test/scala/org/apache/livy/server/interactive/JobApiSpec.scala#L132

I don't know why that is the case but didn't change that existing behavior.

mgaido91 replied:
Because previously the check was in `checkImpersonation`. Since you moved it out from there, you should have been put it somewhere else. The point is: this is not the right place where to put it, since this is not the `proxyUser` which this method should return and to match previous behavior this should be done only when impersonation in enabled (otherwise we'd need None at the end).

risdenk replied:
hmmm yea I see your point. Let me take another look.

risdenk replied:
Commit 3772534 should address this concern.

risdenk commented on server/src/test/scala/org/apache/livy/server/interactive/JobApiSpec.scala:
```diff
@@ -227,3 +226,10 @@ class JobApiSpec extends BaseInteractiveServletSpec {
 
 }
+
+class JobApiSpecNoImpersonation extends JobApiSpec {
```
Currently this shows that the behavior is the same without and without impersonation which is most likely not correct. I need to go back to the master branch and see what the behavior is.

FYI @mgaido91

risdenk replied:
Updated to now have tests that pass on master and fail on this branch. Working on fixing them.

risdenk replied:
ebf3514 fixes these tests.

risdenk commented:
Force pushed only the last commit to add updated tests from master that are now failing on this branch. I need to see how they affect the authorization being pushed around.

risdenk commented:
Latest commit tries to address proxyUser enforcement separate from determining the proxyUser from the request.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -98,23 +99,32 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
-      livyConf: LivyConf): Option[String] = {
-    if (livyConf.getBoolean(LivyConf.IMPERSONATION_ENABLED)) {
-      if (!target.forall(hasSuperAccess(_, requestUser))) {
+      impersonatedUser: String): Boolean = {
```
I am wondering why you are changing this to Boolean and introducing a new method for doing what this was doing earlier. Seems you are just complicating things.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -128,16 +138,16 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
    */
-  def hasModifyAccess(target: String, requestUser: String): Boolean = {
-    requestUser == target || checkModifyPermissions(requestUser)
+  def hasModifyAccess(session: Session, effectiveUser: String): Boolean = {
```
as mentioned, these changes are useless...

mgaido91 commented on server/src/test/scala/org/apache/livy/server/SessionServletSpec.scala:
```diff
@@ -62,27 +59,27 @@ object SessionServletSpec {
-        accessManager.checkImpersonation(params.get(PROXY_USER), remoteUser(req), livyConf)
-        new MockSession(sessionManager.nextId(), remoteUser(req), conf)
+        val bodyImpersonatedUser = params.get(PROXY_USER)
+          .filter(accessManager.checkImpersonation(remoteUser(req), _))
```
I think this should be executed after the orElse of the following line...

risdenk commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -183,11 +207,11 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
   private def doWithSession(fn: (S => Any),
       allowAll: Boolean,
-      checkFn: Option[(String, String) => Boolean]): Any = {
+      checkFn: Option[(Session, String) => Boolean]): Any = {
```
Change left over from checking the session proxy user. Will create separate JIRA for that change.

mgaido91 commented on server/src/main/scala/org/apache/livy/server/SessionServlet.scala:
```diff
@@ -160,6 +160,29 @@ abstract class SessionServlet[S <: Session, R <: RecoveryMetadata](
+   * Gets the request user or impersonated user to determine the effective user.
+   */
+  protected def effectiveUser(request: HttpServletRequest): String = {
+    accessManager.checkImpersonation(impersonatedUser(request), remoteUser(request)).orNull
```
why orNull? Shouldn't be `remoteUser(request)` in that case? I'd expect some NullPointerException to happen with `orNull` or anyway, some problems when checking access permissions (since we have null instead of the requesting user). May you please also add a UT for this?

risdenk replied:
Well so I had a similar question looking at the `checkImpersonation` method. We shouldn't even need to have any `orNull` or anything. We should be able to return a `String` instead of `Option[String]` from `checkImpersonation`. Since we either are impersonating and that succeeds or we are returning `None` from `checkImpersonation` today. 

I have to think about it a bit more and probably won't be until Monday.

risdenk replied:
Ok thought about this a bit more. `remoteUser(request)` I think is correct but there is no guarantee that `remoteUser(request)` will not be null (in the case of no Kerberos security for example). I checked the access checks and null is handled just fine.

I can try to look at adding a unit test for this.

risdenk replied:
JobApiSpecNoImpersonation should handle this case actually. I added some more tests here as well around sending in no user (ie: no authentication enabled where request.getRemoteUser would be null.

mgaido91 commented on server/src/test/scala/org/apache/livy/server/interactive/JobApiSpec.scala:
```diff
@@ -227,3 +230,48 @@ class JobApiSpec extends BaseInteractiveServletSpec {
+class JobApiSpecNoImpersonation extends JobApiSpec {
+  override protected def createConf(): LivyConf = synchronized {
+    super.createConf()
+      .set(LivyConf.IMPERSONATION_ENABLED, false)
```
nit: one the previous line

risdenk replied:
Hm not sure why that was on 2 to start with. Fixed now.

risdenk commented:
There is a `docs/rest-api.md` file? Maybe that is the right place for docs. I could add a section there on the added ability for `doAs` parameter.

risdenk commented:
Last commit made effectiveUser return the remoteUser as a fall back, added tests, and cleaned up style based on review.

mgaido91 commented:
@risdenk yes, that is the right place for sure, but since this parameter is valid for all the endpoints, I am not sure whether to create a new section for common parameters having this single one for the moment or do something else. So I'd like to hear if someone has a good/strong opinion on this.

The code itself now looks good to me. The only point I think still need to be addressed is the documentation.

ajbozarth commented:
I agree that this should be added to `rest-api.md` in some way, but this is the first param shared across all calls so I'm unsure the best way to add it. Given it's optional perhaps giving it it's own section towards the bottom with a quick explanation would be best. How about your give documenting it your best shot and I can give better feedback based on what you write.

risdenk commented on server/src/test/scala/org/apache/livy/server/SessionServletSpec.scala:
```diff
@@ -112,17 +109,60 @@ class SessionServletSpec extends BaseSessionServletSpec[Session, RecoveryMetadat
+        jget[MockSessionView](s"/${res.id}", headers = bobHeaders) { res =>
+          assert(res.owner === ADMIN)
+          assert(res.proxyUser === Some("alice"))
           assert(res.logs === IndexedSeq("log"))
```
Let me double check and I can add a check for the scenario you described as well if it doesn't exist.

risdenk replied:
So the reason this works on line ~165 is because ACLs aren't enabled. In the same `SessionServletSpec.scala` file on line ~245 in the class `AclsEnabledSessionServletSpec` you can see the same test with logs is null. This is because ACLs are enabled.

I don't see the exact scenario you outlined but added a test for admin with `doAs=bob` showing that permissions are adhered to.

risdenk commented:
Added test that @vanzin asked about. Added docs in a new section to document `doAs`. Merged master branch to fix merge conflicts after LIVY-41 was committed to master.

vanzin commented on server/src/test/scala/org/apache/livy/server/interactive/JobApiSpec.scala:
```diff
@@ -227,3 +230,59 @@ class JobApiSpec extends BaseInteractiveServletSpec {
+    }
+  }
+
+  it("should not honor impersonation requests") {
```
I'm not sure I follow these tests. What does it mean to not honor impersonation requests?

I'd expect that if impersonation is disabled and you ask for it, you'd get an error. That makes more sense to me than falling back to some other mode.

risdenk replied:
This matches the behavior from before this change. I added this change on master and made sure the tests pass before pulling to this branch. @mgaido91 was very clear that there should be no behavior changes.

risdenk replied:
To be honest I do agree with you that if impersonation is disabled and you ask for you you should get an error but that is not how Livy was setup previously. You will get an error if impersonation is enabled and you aren't a superuser, but if you have impersonation disabled, it falls back to the request user.

@mgaido91 made it clear not to change the behavior of `AccessManager` `checkImpersonation`

mgaido91 replied:
As I said before, I think that if we want to introduce a change in the behavior we currently have for user impersonation, we should do it in another JIRA/PR.

vanzin commented on server/src/test/scala/org/apache/livy/server/SessionServletSpec.scala:
```diff
@@ -114,17 +111,66 @@ class SessionServletSpec extends BaseSessionServletSpec[Session, RecoveryMetadat
+        jget[MockSessionView](s"/${res.id}", headers = bobHeaders) { res =>
+          assert(res.owner === null)
+          assert(res.proxyUser === None)
+          assert(res.logs === IndexedSeq("log"))
```
Sorry, but this is really bothering me.

The test says "should allow other users to see non-sensitive information", and logs are sensitive information. So if "bob" is not the owner, and is not a super user, then either this test is wrong or the test name is wrong.

Seems like no security is enabled in these tests, so basically everyone has access to everything, so the test name should reflect that.

And if it's true that no security is enabled here, I'm not sure what's the point of adding impersonation-related tests.

risdenk replied:
This test was here before and the behavior is the same as before. The test name has been wrong before this change then.

@mgaido91 asked what happens if you don't send in a username and that is what was added. I also added the `doAs` tests to make sure behavior was the same if you added doAs or not. 

If there are no ACLs then users can do whatever they want. This is true before this doAs was added too.

I can update the test name but I want to be clear this was like this before `doAs`.

mgaido91 replied:
My point is: the focus of this PR is adding the parameter `doAs` in order to allow user impersonation. The authorization logic is orthogonal to this change. Adding tests for this situation I think is helpful in order to determine and ensure the current behavior, since they were missing.

If we decide that the logic/behavior we have currently for the authorization, I think we can change it adding some parameters in order to ensure users who upgrade can have no behavior change. But anyway, I think this should not be done in this PR, since the scope of this PR is to add the `doAs` parameter, not to change the impersonation and access control logic.

vanzin replied:
I'm fine with the "behavior changes in another PR" argument. But the test name here is so obviously wrong, that if someone else ends up touching this, they'll be scratching their heads.

So at the very least, the test name must be fixed.

risdenk replied:
Updated the test name.

risdenk commented on server/src/main/scala/org/apache/livy/server/AccessManager.scala:
```diff
@@ -107,9 +107,8 @@ private[livy] class AccessManager(conf: LivyConf) extends Logging {
    */
   def checkImpersonation(
```
@vanzin This is the method that is causing most of your concerns I think. It didn't change with `doAs`. Basically if impersonation is disabled, then it just falls through with no error.

vanzin commented on thriftserver/server/src/main/scala/org/apache/livy/thriftserver/LivyThriftSessionManager.scala:
```diff
@@ -230,6 +230,7 @@ class LivyThriftSessionManager(val server: LivyThriftServer, val livyConf: LivyC
         server.livySessionManager.nextId(),
         None,
         username,
+        None,
```
Hmm, looks like the thrift side doesn't yet handle impersonation, so this is fine. Would probably be good to have a bug for that.

mgaido91 replied:
it does handle it but internally, as Hive server does, not using Livy impersonation (please see https://github.com/apache/incubator-livy/blob/47d3ee6b6555a63d0b871788a53aab022fff518a/thriftserver/server/src/main/scala/org/apache/livy/thriftserver/cli/ThriftCLIService.scala#L265)

vanzin replied:
Interesting; but at least now, it doesn't seem like that information is being reflected in Livy's state. Shouldn't matter much, but it's something to keep in mind.

vanzin commented:
Merging to master.
